### PR TITLE
V2: Refactor tests for data split

### PR DESCRIPTION
**Pull request by oscarwumit**
## Description
Refactored tests for data split for chemprop v2.

### CODE REVIEW

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (JacksonBurns) [inline chemprop/v2/tests/unit/data/test_data_utils.py:24]: This specification is actually allowed in `astartes`, which will raise a `NormalizationWarning` and then normalize the inputs to 1. I am going to modify this test to check that the warning is properly raised.

**Review by JacksonBurns — COMMENTED** [review]
Thanks for the refactor - I am adding a unit test (and fixing) the scaffold splitting.

**Review by JacksonBurns — COMMENTED** [review]
This PR now includes some bug fixes in `split_data` in addition to the improved unit tests. @oscarwumit thanks for getting this started, let's get another review on it.

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (davidegraff) [inline chemprop/v2/tests/unit/data/test_data_utils.py:28]:
```suggestion
@pytest.fixture(params=[["C", "CC", "CCC", "C1CC1", "C1CCC1"]])
def molecule_dataset_with_rings(request):
    """A dataset with rings (for scaffold splitting)"""
    return [MoleculeDatapoint.from_smi(s) for s in request.param]
```
also change the above `fixture`s

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (davidegraff) [inline chemprop/v2/tests/unit/data/test_data_utils.py:40]: same question as above

**Review by davidegraff — COMMENTED** [review]
I guess also a general comment on the content of these tests: many of them fix the random seed and then define the expected split composition. Will these tests still be accurate if in a future version the underlying RNG of python/numpy changes?

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (JacksonBurns) [reply chemprop/v2/tests/unit/data/test_data_utils.py:34]: astartes raises this, it's import from astartes.utils.warnings

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (JacksonBurns) [reply chemprop/v2/tests/unit/data/test_data_utils.py:28]: @oscarwumit your call on this

**Comment by JacksonBurns** [comment]
> I guess also a general comment on the content of these tests: many of them fix the random seed and then define the expected split composition. Will these tests still be accurate if in a future version the underlying RNG of python/numpy changes?

No, if the RNG in numpy changes these tests will fail. That will be important for us, though, since we would want to warn our users that numpy version 1.x produces different splits with the same random seed see than version 1.x-1 or whatever it may be. I doubt this will actually happen without the numpy devs publishing extensive warnings beforehand, but better to be safe than sorry.

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (davidegraff) [reply chemprop/v2/tests/unit/data/test_data_utils.py:34]: then we shouldn't be testing this as it's actually just testing `astartes` code not our own client logic

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (davidegraff) [reply chemprop/v2/tests/unit/data/test_data_utils.py:40]: should move from `assert` to an appropriate `Exception` then

**Comment by davidegraff** [comment]
> > I guess also a general comment on the content of these tests: many of them fix the random seed and then define the expected split composition. Will these tests still be accurate if in a future version the underlying RNG of python/numpy changes?
> 
> No, if the RNG in numpy changes these tests will fail. That will be important for us, though, since we would want to warn our users that numpy version 1.x produces different splits with the same random seed see than version 1.x-1 or whatever it may be. I doubt this will actually happen without the numpy devs publishing extensive warnings beforehand, but better to be safe than sorry.

Why would we need a test then if we know that a change to RNG internals will no longer guarantee reproducibility? It could well be that some of our tests still pass with a hypothetical RNG change, but that still wouldn't mean we can guarantee reproducibility because our tests cover only an infinitesimally small fraction of the space of possible splits. IMO what we want to be testing here is:
1. **repeatability**: If we call `split_data` several times and reset the seed to the same value in between each call, can we generate the same splits? We should be able to knowingly "control" the randomness of our data splitting, but if we fail these tests, then it indicates we're not actually doing so correctly.
2. **faithfulness**: does `split_data` generate the same splits if we were to manually generate splits using `astartes` via the same strategy? That is, we _expect_ `split_data` to faithfullly generate the results as `astartes` (while maintaining modularity and hiding details) but is it actually equivalent?

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (JacksonBurns) [reply chemprop/v2/tests/unit/data/test_data_utils.py:34]: I was more validating that we don't bury the warning - seems like something we would want to make it all the way to the user

**Comment by oscarwumit** [comment]
@davidegraff @JacksonBurns Many thanks for your feedback. I have refactored the code to reflect our discussion. In the new version:
1.  I modified the way we test random seed. Now it compares the expected result with what is generated from the desired data splitting backend in a way that is less dependent on a hypothetical RNG change.
2. I kept and modified some of the previous tests on different splitting methods using explicit smiles outcomes with the following consideration:
- these tests first make sure that a specific splitting type requested by user runs properly
- in case of failure, error message is added to alert the user that the corresponding data splitting backend might have changed and needs attention 
I agree with the discussion that it is largely not our duty to control the behaviour of any underlying RNG or backend. However, tests that explicitly compares to expected smiles can serve as a useful alarm to the user and developer that the behaviour of some critical backend may have changed. This could make debugging easier. Also, I added message that clearly states that the failure of such tests does not necessarily mean chemprop is broken. It is more likely that the backend have changed. 

Let me know if these fixes address your concerns, and thanks again for your insightful review.

File: chemprop/v2/data/utils.py
Comment (davidegraff) [inline chemprop/v2/data/utils.py:63]: `raise` a `ValueError`--no `assert`s in the main code

File: chemprop/v2/data/utils.py
Comment (davidegraff) [inline chemprop/v2/data/utils.py:182]: same here:
```suggestion
        *_, train_idxs, test_idxs = result
```

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (davidegraff) [reply chemprop/v2/tests/unit/data/test_data_utils.py:36]: yeah but the point is here there's nothing that this code is actually testing in the `split_data` function. There's no logic in our client code that does _anything_ when the splits are not normalized properly, so therefore there's nothing to test. If we were explicitly catching the warning and re-raising it, then this test would be appropriate.

File: chemprop/v2/data/utils.py
Comment (JacksonBurns) [reply chemprop/v2/data/utils.py:180]: Prefer to be explicit here, though this is cool

**Comment by JacksonBurns** [comment]
We're going to hold on the test asserting that a normalization warning is raised to keep ourselves responsible for proper loogging. We're ensuring that we _don't_ catch and recast, or catch and ignore, at any point in the future.

**Comment by davidegraff** [comment]
> We're going to hold on the test asserting that a normalization warning is raised to keep ourselves responsible for proper logging. We're ensuring that we _don't_ catch and recast, or catch and ignore, at any point in the future.

Why would we not allow that? Presumably that's an informed decision made by a future developer that might cause this test to fail. E.g., if they decide that they'd rather raise a `ValueError` earlier in `split_data()`, then this test will fail. Would that be a true failure? IMO it's the test that's wrong in this case. Also, the logical extension of this argument is that we should be testing all of our PyTorch code that we're correctly passing through `RuntimeError`s every time we pass an improperly shaped tensor. But of course, by virtue of the fact that our code does _nothing_ (which we know when writing tests because we have access to the source), we know that these tests will pass. Again, if we decide at a later date to explicitly catch these `RuntimeError`s and rethrow them as `CustomError`s, these tests will fail and not for a valid reason.

File: chemprop/v2/data/utils.py
Comment (davidegraff) [reply chemprop/v2/data/utils.py:181]: this is being explicit, just in a way you're not familiar with. This explicit use-case is why the feature was accepted as PEP over a decade ago.

**Comment by JacksonBurns** [comment]
This future developer should know that it was an intentional design choice that this warning should make it to the user, and this unit test asserts that.

I think it's a slippery slope fallacy to argue that including this test justifies unit testing all runtime errors. It's much more forseeable that a future developer would say "this warning is annoying, let's catch and ignore it" rather than "this runtime error keeps stopping execution, lets catch it and try and move on anyway".

**Comment by davidegraff** [comment]
> This future developer should know that it was an intentional design choice that this warning should make it to the user, and this unit test asserts that.

But all warnings/exceptions do get propagated _by default_, and the choice to do nothing **does not** mean that this warning being raised is expected behavior. If we actually want it to be a feature of this function, then we should be _explicitly_ raising it from inside the function rather than _implicitly_ allowing it to propagate. This also goes back to the broader point of that this test is _not testing chemprop_. If at a future point, `astartes` changes its internals to raise a different warning or raise an exception, then our test will fail but not due to **any** fault in our client code. Rather, it will fail because we're testing code that is not our own and thus we have no control over it (ignoring the fact that you're the `astartes` dev).
 
> I think it's a slippery slope fallacy to argue that including this test justifies unit testing all runtime errors. It's much more forseeable that a future developer would say "this warning is annoying, let's catch and ignore it" rather than "this runtime error keeps stopping execution, lets catch it and try and move on anyway".

Even if you want to argue on the basis of logical fallacies, the point still stands. We should be consistent in how we write our code and how we test it. Limiting the discussion to just `Warning`s, there are a number of warnings that get raised from the `lightning` library that propagate because we've chosen to do nothing (as in `split_data()`). Should we also be testing those?

File: chemprop/v2/tests/unit/data/test_data_utils.py
Comment (kevingreenman) [reply chemprop/v2/tests/unit/data/test_data_utils.py:38]: I agree with this change. @oscarwumit please update this and the other tests where we currently have `smiles_list` instead of specifying `params` in the `@pytest.fixture`

**Comment by kevingreenman** [comment]
> It's extremely uncommon to provide messages with `assert`s in testing. If you want to explain the test, then this message should go in the docstring. Most testing frameworks will show this docstring if you increase the verbosity level. If you want to add a very detailed, test-specific failure message, then I might be able to see the value. But if it's the same generic failure message, then it should go in the docstring IMO

@oscarwumit I agree with this, please put the messages in the docstrings.

**Review by kevingreenman — CHANGES_REQUESTED** [review]
Thanks to @oscarwumit  and @JacksonBurns for this PR and to @davidegraff  for the reviews so far. I'm going to attempt to find some middle ground here. In the interest of getting this PR merged soon so we can move on to next steps, in places where I'm indifferent or think both arguments are reasonable, my decision will default to the option that results in fewer changes to the PR.

I generally agree with @davidegraff that we should be testing our own code rather than underlying dependencies/backends. However, given the importance of splitting to the functioning of our code, I think it's okay to keep the tests that @oscarwumit has written with explicit SMILES outcomes (in addition to the tests that he wrote to compare our wrappers to the raw astartes outputs). @JacksonBurns are the same / similar tests for explicit SMILES outputs already implemented within astartes as well?

In the case of the `NormalizationWarning`, I think that although it is meant to test an intentional design choice / keep ourselves accountable in logging, this is an arbitrary decision to keep ourselves accountable for this and not other warnings. However, the code is already written, so I am okay with leaving it in for now.

I left a couple other comments about `params` in `pytest.fixtures` and moving messages to docstrings in some tests. After these changes are made, I will approve.

**Comment by JacksonBurns** [comment]
> @JacksonBurns are the same / similar tests for explicit SMILES outputs already implemented within astartes as well?

`astartes` does explicitly check certain SMILES splitting outcomes for all of the samplers we have here.

> I left a couple other comments about `params` in `pytest.fixtures` and moving messages to docstrings in some tests. After these changes are made, I will approve.

Done in fe0b9ee

**Review by kevingreenman — CHANGES_REQUESTED** [review]
> > It's extremely uncommon to provide messages with `assert`s in testing. If you want to explain the test, then this message should go in the docstring. Most testing frameworks will show this docstring if you increase the verbosity level. If you want to add a very detailed, test-specific failure message, then I might be able to see the value. But if it's the same generic failure message, then it should go in the docstring IMO
> 
> @oscarwumit I agree with this, please put the messages in the docstrings.

@JacksonBurns @oscarwumit still needs this final change

**Comment by JacksonBurns** [comment]
Forgot about that my apologies, done in 9bb2c02

**Review by kevingreenman — APPROVED** [review]

**Comment by kevingreenman** [comment]
great work, thanks @JacksonBurns !

**Pull request merged**
